src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -605,18 +715,27 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
             Compilation currentCompilation,
             ArrayBuilder<TypeParameterDiagnosticInfo> diagnosticsBuilder,
             ref ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder,
+            ref ArrayBuilder<DiagnosticInfo> tupleDiagnosticsBuilder,
             BitVector skipParameters = default(BitVector))
         {
-            return CheckConstraints(
-                method,
-                conversions,
-                method.TypeSubstitution,
-                ((MethodSymbol)method.OriginalDefinition).TypeParameters,
-                method.TypeArguments,
-                currentCompilation,
-                diagnosticsBuilder,
-                ref useSiteDiagnosticsBuilder,
-                skipParameters);
+            bool result = CheckConstraints(
+                            method,
+                            conversions,
+                            method.TypeSubstitution,
+                            ((MethodSymbol)method.OriginalDefinition).TypeParameters,
+                            method.TypeArguments,
+                            currentCompilation,
+                            diagnosticsBuilder,
+                            ref useSiteDiagnosticsBuilder,
+                            skipParameters);
+
+            foreach (var parameter in method.Parameters)
+            {
+                result &= CheckTupleNamesAgainstBase(parameter.Type, ref tupleDiagnosticsBuilder);
```

Comment:
Notes from our discussion:
I'll introduce a number of shortcuts to make those checks cheaper:
- Visit and check for tuples, if none then cut short
- when checking types (as opposed to methods) no deep check for the top-level type (only for the base types)
- stop going up bases if they are not "variable" (they don't use type parameters)

On the other hand, checking methods (including during overload resolution) needs to do a deep check.